crates/common_utils/src/types/keymanager.rs

```diff
@@ -39,3 +58,425 @@ pub struct DataKeyCreateResponse {
+}
+
+impl From<(FxHashMap<String, Encryption>, Identifier)> for BatchDecryptDataRequest {
+    fn from((map, identifier): (FxHashMap<String, Encryption>, Identifier)) -> Self {
+        let group = map
+            .into_iter()
+            .map(|(k, v)| {
+                (
+                    k,
+                    EncryptedData {
+                        data: StrongSecret::new(v.clone().into_inner().expose()),
+                    },
+                )
+            })
+            .collect();
+        Self {
+            data: EncryptedDataGroup(group),
+            identifier,
+        }
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct BatchDecryptDataResponse {
+    pub data: DecryptedDataGroup,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct DecryptDataResponse {
+    pub data: DecryptedData,
+}
+
+#[derive(Clone, Debug)]
+pub struct DecryptedData(StrongSecret<Vec<u8>>);
+
+impl Serialize for DecryptedData {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let data = BASE64_ENGINE.encode(self.0.peek());
+        serializer.serialize_str(&data)
+    }
+}
+
+impl<'de> Deserialize<'de> for DecryptedData {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct DecryptedDataVisitor;
+
+        impl<'de> Visitor<'de> for DecryptedDataVisitor {
+            type Value = DecryptedData;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+                formatter.write_str("string of the format {version}:{base64_encoded_data}'")
+            }
+
+            fn visit_str<E>(self, value: &str) -> Result<DecryptedData, E>
+            where
+                E: de::Error,
+            {
+                let dec_data = BASE64_ENGINE.decode(value).map_err(|err| {
+                    let err = err.to_string();
+                    E::invalid_value(Unexpected::Str(value), &err.as_str())
+                })?;
+
+                Ok(DecryptedData(dec_data.into()))
+            }
+        }
+
+        deserializer.deserialize_str(DecryptedDataVisitor)
+    }
+}
+
+impl DecryptedData {
+    pub fn from_data(data: StrongSecret<Vec<u8>>) -> Self {
+        Self(data)
+    }
+
+    pub fn inner(self) -> StrongSecret<Vec<u8>> {
+        self.0
+    }
+}
+
+#[derive(Debug)]
+pub struct EncryptedData {
+    pub data: StrongSecret<Vec<u8>>,
+}
+
+impl Serialize for EncryptedData {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let data = match String::from_utf8(self.data.peek().clone()) {
+            Ok(data) => data,
+            Err(_) => {
+                let data = BASE64_ENGINE.encode(self.data.peek().clone());
+                format!("{DEFAULT_ENCRYPTION_VERSION}:{data}")
+            }
+        };
+        serializer.serialize_str(data.as_str())
+    }
+}
```

Comment:
I wouldn't prefer having such fallback logic as part of `Serialize` / `Deserialize` implementations, I see this as business logic that is leaking into the serialization implementation. Can we instead move this to a module that we can use with `#[serde(with = "module_name")]`?